Refactor lookup table helper and apply CRIS v28 changes

johnclary opened the pull request:
## Associated issues

- https://github.com/cityofaustin/atd-data-tech/issues/18938
- https://github.com/cityofaustin/atd-data-tech/issues/18826
- https://github.com/cityofaustin/atd-data-tech/issues/18937

## Testing

**URL to test:** Local

You're going to want to test this PR by applying the migrations incrementally. First you will test the lookup table constraint changes, then you can test the helper script.

1. Replicate prod and start your stack. Apply the first four migrations in this PR

```shell
$ hasura migrate apply --up 4
```

2. Now apply metadata, which will test that the foreign key relationships we're expecting are still intact:

```shell
$ hasura metadata apply
```

Nice! Now you can test the helper script. 

3. The script readme will walk you through the steps to test this helper.  [Here](https://github.com/user-attachments/files/16852420/lookup_current.xml.txt) is a copy of the file that you can save as `lookups.xml` in the `./cris_data` directory of the new toolbox.

4. Once you've run the helper script, you can inspect the up and down migrations that have been created. Unfortunately, you will not be able to apply those migrations without making the modifications I made in the `1725561923206_cris_v28_lookup_changes` migration.

5. To apply the remaining migration, run `hasura migrate apply`.

6.  Kick the tires on the VZE and make sure it's operating normally.

7. Finally, run the down migrations: `hasura migrate apply --down 5`

---
#### Ship list
- [ ] Check migrations for any conflicts with latest migrations in master branch
- [ ] Confirm Hasura role permissions for necessary access
- [ ] Code reviewed
- [ ] Product manager approved

johnclary commented on atd-toolbox/get_lookup_table_changes/README.md:
i added a note about this script to the generic `./atd-toolbox` README in [this follow-up PR: https://github.com/cityofaustin/atd-vz-data/pull/1543](https://github.com/cityofaustin/atd-vz-data/pull/1543)

frankhereford commented on atd-toolbox/get_lookup_table_changes/docs/extract_config_2.png:
Do you think it would be helpful to include all three counties in this diagram? I pulled up the crashes in qgis just to double check -- forgive the ugliest map ever created.

![image](https://github.com/user-attachments/assets/52909ae2-bb26-4b96-bc9f-846bd5b89d2f)

johnclary replied:
yes, great point. when i made these instructions i was thinking "steps to get the smallest/quickest extract possible", because we just need that lookups.xml file. but i don't want it sitting out there that this might be the correct way to get a CRIS extract for the standard import.

we have docs elsehwere (that need to be updated) with the steps to request an extract, so i will update this readme to point to those instead 🙏

frankhereford commented on atd-vzd/migrations/default/1725561923206_cris_v28_lookup_changes/up.sql:
```diff
@@ -0,0 +1,105 @@
+delete from lookups.agency where id = 3323;
+delete from lookups.agency where id = 3335;
+delete from lookups.agency where id = 42;
+delete from lookups.agency where id = 71;
```
The crashes with `crashes.id` of the following use this `lookups.agency` record and the migration fails when run on today's replicated data.


```
334511, 337719, 340347, 344033, 345331, 347562, 356144, 401431, 404437, 404442
```

mddilley replied:
i saw the same error when running the up migration after replicating data today.

johnclary replied:
yes! the migration failed by design, but the conflict is surprise to me 😞  thank you both for testing. i'm looking to this.

frankhereford commented on atd-vzd/migrations/default/1725474997305_rebuild-lookup-table-constraints/up.sql:
There's nothing wrong with this file, but I want to observe that it takes a bit to run on my mac, so we should be thinking about the 30 second timeout on the HTTP session between the action issuing the migration and our DB endpoint.

frankhereford replied:
```
(venv) frank@Franks-MacBook-Pro-5 database % time hasura migrate apply --up 1                                       (merge-in-other-prs|✚1⚑7)
? Select a database to use All (all available databases)
INFO migrations applied on database: default                                                                                                  
hasura migrate apply --up 1  0.96s user 0.63s system 8% cpu 17.717 total
```

~ 17 seconds on my m1

johnclary replied:
hmm, you're right—that's cutting it a bit close. i'll try splitting this into a few smaller pieces, just to be safe.

frankhereford commented on atd-vzd/migrations/default/1725561923206_cris_v28_lookup_changes/up.sql:
After I comment out the lines with the FK issues below, just to make it work, it does in fact work, but takes 16 seconds to run on my mac. Another 🐌 slow one that's doing just exactly what it should and takes a reasonable time -- we should consider what the migration actions are going to look like here too.

mddilley commented on atd-toolbox/get_lookup_table_changes/README.md:
````diff
@@ -1,27 +1,72 @@
 
-`pip install -r requirements.txt`
+```shell
+$ pip install -r requirements.txt
````
I set up a fresh venv when testing, and I noticed that `requests` is missing from the requirements file.

roseeichelmann replied:
:pray:

johnclary replied:
thanks for catching this, Mike!

mddilley reviewed (commented):
I left a question about what we should expect when we run into a restricted delete. The docs for getting the xml look great, and thanks for the cleanup too!

johnclary commented:
OK, so digging a bit more into the issue with the `agency_id` deletes, i see that IDs `71` and `2296` are the problem, and i see that the small handful of crashes using these IDs are dated before 2012. Which to to say, they're very old crash records.

I think what may be going on is that, in general, CRIS does update/migrate crash records so that they meet the latest schema/release version, but these records were outside of the 10-year cutoff, and so they were not updated/we did not receive any updates to make them compliant with the new schema.

In any case, I think we have just two options:
1. Migrate these old records to the correct agency values by hand
2. Purge these old records

I will do option 1 but this is a compelling case for not keeping old records around in our primary DB tables.

johnclary commented:
Requesting review after having done these things:
- Split the slow part of the migrations into two. Hopefully this gives us enough head room for the prod deploy.
- Add special handling for the breaking changes to `lookups.agency`
- Update the readme to mention this unfortunate scenario

chiaberry commented:
I was following the test steps exactly and then realized that the code has changed since the instructions were written, which is why i was getting metadata inconsistencies

```console
hasura migrate apply --up 4 
```

chiaberry commented on atd-toolbox/get_lookup_table_changes/README.md:
````diff
@@ -65,7 +65,9 @@ $ hasura migrate create lookup_table_migrations_cris_v28
 ```
 
-5. Finally, re-apply metadata to check for errors. 
+In some cases, the migrations applied by this helper may fail to due to conflicts with historical records. In those cases, you will need to manually create/edit the migrations and use your best judegment as to how to approach the change. PR [#1541](https://github.com/cityofaustin/atd-vz-data/pull/1541) covers an example of this scenario.
````
tiny typo in judgement

chiaberry reviewed (approved):
ran through the steps, encountered the expected error with the old records. nice to see all the cleanup too

johnclary commented:
I'm going to retest this momentarily—there should not be any errors 🫠

johnclary commented:
OK, I updated the test steps to mention that you cannot copy/paste the auto-generated migrations because they need to modified by hand. I updated the test steps, ran through them again, and everything is working as expected 🙏

roseeichelmann commented on atd-toolbox/README.md:
```diff
@@ -2,6 +2,32 @@
+
+Various scripts used as part of launching the new data model for VZ v2.0. We want to keep these around for the few months after launch (August 2024) but they can be archived in the near future.
+
+### Engineering Areas and council districts load tools - `engineering_areas_load`, `load_council_districts`
```
thanks for adding these script descriptions :)) this README got forgotten about with time methinks

roseeichelmann commented on atd-toolbox/get_lookup_table_changes/README.md:
```diff
@@ -1,27 +1,74 @@
+
+![CRIS extract config - page 2](docs/extract_config_2.png) 
+
+![CRIS extract config - page 3](docs/extract_config_3.png)
```
super helpful README instructions, thank you!!

roseeichelmann commented on atd-toolbox/get_lookup_table_changes/get_lookup_table_changes.py:
```diff
@@ -1,258 +1,154 @@
+                continue
+            # if this happens we'll have to figure out how to deal with this situation
+            raise Exception(
+                """It appears a lookup table has been deleted or renames by CRIS. This script cannot handle that case. Aborting."""
```
small type here "renames"

roseeichelmann commented on atd-toolbox/get_lookup_table_changes/get_lookup_table_changes.py:
```diff
@@ -1,258 +1,154 @@
-            # Now check for any values in our lookup tables that aren't in the extract and are possibly crufty
-            for key in our_table_dict:
-                if key not in extract_table_dict:
+        print(f"Gettling lookup values for {table_name}")
```
same typo "gettling"

roseeichelmann commented on atd-vzd/migrations/default/1725561923206_cris_v28_lookup_changes/up.sql:
```diff
@@ -0,0 +1,105 @@
+delete from lookups.agency where id = 2202;
+delete from lookups.agency where id = 220;
+delete from lookups.agency where id = 2272;
+delete from lookups.agency where id = 2296;
```
it is so awesome that the data model is catching upstream errors now!

roseeichelmann reviewed (approved):
just read through and tested all this code, it looks so great! not getting any errors anymore after your updates. love the updates you made to `get_lookup_table_changes`. this is :rocket: :ship:

johnclary commented:
Thank you!
